src/Build/BuildCheck/Analyzers/NoEnvironmentVariablePropertyAnalyzer.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Build.Construction;
+
+namespace Microsoft.Build.Experimental.BuildCheck.Analyzers;
+
+internal sealed class NoEnvironmentVariablePropertyAnalyzer : BuildAnalyzer
+{
+    /// <summary>
+    /// Contains the list of reported environment variables.
+    /// </summary>
+    private readonly IList<EnvironmentVariableIdentityKey> _environmentVariablesReported = new List<EnvironmentVariableIdentityKey>();
+
+    public static BuildAnalyzerRule SupportedRule = new BuildAnalyzerRule(
+        "BC0103",
```

Comment:
Let's add mention here as well: https://github.com/dotnet/msbuild/blob/main/documentation/specs/proposed/BuildCheck/Codes.md

And let's mention in the description why it is considered a bad practice (it hurts reproducibility and isolate-ability of the builds)